src/sentry/models/projectsdk.py

```
@@ -0,0 +1,162 @@
+        return (
+            (cls.PROFILE.value, "profile"),
+            (cls.PROFILE_CHUNK.value, "profile_chunk"),
+        )
+
+
+@region_silo_model
+class ProjectSDK(DefaultFieldsModel):
+    __relocation_scope__ = RelocationScope.Organization
+
+    date_updated = models.DateTimeField(auto_now=True)
+
+    project = FlexibleForeignKey("sentry.Project", on_delete=models.CASCADE)
+    event_type = BoundedIntegerField(choices=EventType.as_choices())
+    sdk_name = models.CharField()
+    sdk_version = models.CharField()
+
+    class Meta:
+        unique_together = (("project", "event_type", "sdk_name"),)
+
+    __repr__ = sane_repr("project", "event_type", "sdk_name", "sdk_version")
+
+    @classmethod
+    def get_lock_key(cls, project: Project, event_type: EventType, sdk_name: str):
+        return f"lprojectsdk:{project.id}:{event_type.value}:{md5_text(sdk_name).hexdigest()}"
+
+    @classmethod
+    def get_cache_key(cls, project: Project, event_type: EventType, sdk_name: str):
+        return f"projectsdk:{project.id}:{event_type.value}:{md5_text(sdk_name).hexdigest()}"
+
+    @classmethod
+    def update_with_newest_version_or_create(
+        cls,
+        project: Project,
+        event_type: EventType,
+        sdk_name: str,
+        sdk_version: str,
+    ):
+        try:
+            new_version = parse_version(sdk_version)
+        except InvalidVersion:
+            # non-semver sdk version, ignore and move on
+            return
+
+        normalized_sdk_name = normalize_sdk_name(sdk_name)
+        if normalized_sdk_name is None:
+            logger.info("Unknown sdk name: %s", sdk_name)
+            return
+
+        lock_key = cls.get_lock_key(project, event_type, normalized_sdk_name)
+        lock = locks.get(lock_key, duration=10, name="projectsdk")
+
+        # This can raise `sentry.utils.locking.UnableToAcquireLock`
+        # that needs to be handled by the caller and handled
+        # appropriately with retries if needed.
```

You could also just catch it in here, but I guess maybe that behaviour is a bit unclear and less flexible

---

Yeah, I thought about catching it here but that means this method can now silently fail and not update anything which seems like a foot gun.